storm-core/src/jvm/org/apache/storm/daemon/supervisor/ContainerLauncher.java

```
@@ -0,0 +1,103 @@
+        if (ConfigUtils.isLocalMode(conf)) {
+            return new LocalContainerLauncher(conf, supervisorId, sharedContext);
+        }
+        
+        ResourceIsolationInterface resourceIsolationManager = null;
+        if (Utils.getBoolean(conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN_ENABLE), false)) {
+            resourceIsolationManager = Utils.newInstance((String) conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN));
+            resourceIsolationManager.prepare(conf);
+            LOG.info("Using resource isolation plugin {} {}", conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN), resourceIsolationManager);
+        }
+
+        if (Utils.getBoolean(conf.get(Config.SUPERVISOR_RUN_WORKER_AS_USER), false)) {
+            return new RunAsUserContainerLauncher(conf, supervisorId, resourceIsolationManager);
+        }
+        return new BasicContainerLauncher(conf, supervisorId, resourceIsolationManager);
+    }
+    
+    protected ContainerLauncher() {
+        //Empty
+    }
+
+    /**
+     * Launch a container in a given slot
+     * @param port the port to run this on
+     * @param assignment what to launch
+     * @param state the current state of the supervisor
+     * @return The container that can be used to manager the processes.
+     * @throws IOException on any error 
+     */
+    public abstract Container launchContainer(int port, LocalAssignment assignment, LocalState state) throws IOException;
+    
+    /**
+     * Recover a container for a running process
+     * @param port the port the assignment is running on
+     * @param assignment the assignment that was launched
+     * @param state the current state of the supervisor
+     * @return The container that can be used to manage the processes.
+     * @throws IOException on any error
+     */
+    public abstract Container recoverContainer(int port, LocalAssignment assignment, LocalState state) throws IOException;
```

Yes I have been meaning to switch it over to throw an exception instead of returning null.  Will do that